File: 25. Reverse Nodes in k-Group/25. Reverse Nodes in k-Group.cpp
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int& k) {
        if(!head)   return head;

        ListNode* prev = NULL, *curr = head, *temp = head;
        for(int i = 0; i < k; ++i){
            if(!temp)
                return head;
            temp = temp->next;
        }

        for(int i = 0; i < k; ++i){
            temp = curr->next, curr->next = prev, prev = curr, curr = temp;
        }

        head->next = reverseKGroup(curr, k);
        return prev;
    }
};

File: 25. Reverse Nodes in k-Group/25. Reverse Nodes in k-Group.java
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode newHead = head, last = head, temp, prev = null, ans;
        int count = 0;
        while(newHead != null){
            newHead = newHead.next;
            ++count;
        }
        for(int j = 0; j < k; ++j){
            temp = head.next;
            head.next = prev;
            prev = head;
            head = temp;
        }
        ans = prev;
        for(int i = count/k; i > 1; --i){
            newHead = head;
            for(int j = 0; j < k; ++j){
                temp = head.next;
                head.next = prev;
                prev = head;
                head = temp;
            }
            last.next = prev;
            last = newHead;
        }
        last.next = head;
        return ans;
    }
}